hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/SqlPageCodec.java

```diff
@@ -0,0 +1,181 @@
+
+                    break;
+
+                case BOOLEAN:
+                    ListCNBooleanCodec.encode(clientMessage, (Iterable<Boolean>) column);
+
+                    break;
+
+                case TINYINT:
+                    ListCNByteCodec.encode(clientMessage, (Iterable<Byte>) column);
+
+                    break;
+
+                case INTEGER:
+                    ListCNIntegerCodec.encode(clientMessage, (Iterable<Integer>) column);
+
+                    break;
+
+                case BIGINT:
+                    ListCNLongCodec.encode(clientMessage, (Iterable<Long>) column);
+
+                    break;
+
+                case SMALLINT:
+                case DECIMAL:
+                case REAL:
+                case DOUBLE:
+                case DATE:
+                case TIME:
+                case TIMESTAMP:
+                case TIMESTAMP_WITH_TIME_ZONE:
+                case NULL:
+                case OBJECT:
+                    assert SqlPage.convertToData(columnType);
+
+                    ListMultiFrameCodec.encode(clientMessage, (Iterable<Data>) column, DataCodec::encodeNullable);
+
+                    break;
+
+                default:
+                    throw new IllegalStateException("Unknown type " + columnType);
+            }
+        }
+
+        clientMessage.add(END_FRAME.copy());
+    }
+
+    @SuppressWarnings("checkstyle:CyclomaticComplexity")
+    public static SqlPage decode(ClientMessage.ForwardFrameIterator iterator) {
+        // begin frame
+        iterator.next();
+
+        // Read the "last" flag.
+        boolean isLast = iterator.next().content[0] == 1;
+
+        // Read column types.
+        List<Integer> columnTypeIds = ListIntegerCodec.decode(iterator);
+        List<SqlColumnType> columnTypes = new ArrayList<>(columnTypeIds.size());
+
+        // Read columns.
+        List<List<?>> columns = new ArrayList<>(columnTypeIds.size());
+
+        for (int columnTypeId : columnTypeIds) {
+            SqlColumnType columnType = SqlColumnType.getById(columnTypeId);
+            assert columnType != null;
+
+            columnTypes.add(columnType);
+
+            switch (columnType) {
+                case VARCHAR:
+                    columns.add(ListMultiFrameCodec.decodeContainsNullable(iterator, StringCodec::decode));
+
+                    break;
+
+                case BOOLEAN:
+                    columns.add(ListCNBooleanCodec.decode(iterator));
+
+                    break;
+
+                case TINYINT:
+                    columns.add(ListCNByteCodec.decode(iterator));
+
+                    break;
+
+                case INTEGER:
+                    columns.add(ListCNIntegerCodec.decode(iterator));
+
+                    break;
+
+                case BIGINT:
+                    columns.add(ListCNLongCodec.decode(iterator));
+
+                    break;
+
+                case SMALLINT:
```

Comment:
Some of them should be straightforward:

SmallInt -> Bits.writeShortL
Real-> writeInt(Float.floatToIntBits(v))
Double -> writeLong(Double.doubleToLongBits(v));
null-> Don't write anything and reutrn null inresposne

Reply:
Agree. But the idea was to merge the infrastructure first and then switch to other types. This way we will have PRs with a smaller scope that is easier to reason about. If we attempt to merge many new types (especially the base types), it may raise multiple concerns that will delay the merge. Does it make sense?

Reply:
Yes, make sense. I just wanted to share how I think we should serialize these and these are much more straighforward to me than the new types(Time related and decimal).